### PR TITLE
Support reactor timing on more reactors.

### PR DESCRIPTION
**Pull request by clokep**
Supports timing reactor ticks on `select` and `poll` reactors, as well as `epoll` (which we previously supported). Also supports reactor timings when using the `asyncio` reactor (when used with an underlying `select`, `poll`, `epoll`, `/dev/poll`, or `kqueue` event loop).

### CODE REVIEW

**Comment by DMRobertson** [comment]
This seems unhappy on Linux (GHA runners, anyway):

```
  File "/home/runner/work/synapse/synapse/synapse/metrics/__init__.py", line 47, in <module>
    import synapse.metrics._reactor_metrics  # noqa: F401
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/synapse/synapse/synapse/metrics/_reactor_metrics.py", line 122, in <module>
    wrapper = reactor._poller.poll = CallWrapper(reactor._poller.poll)
              ^^^^^^^^^^^^^^^^^^^^
AttributeError: 'select.epoll' object attribute 'poll' is read-only
```

**Comment by clokep** [comment]
> This seems unhappy on Linux (GHA runners, anyway):

I was unable to fully test on Linux locally, was waiting for builds to run.

**Comment by DMRobertson** [comment]
Roger. Shout if you want a test minion to try anything out!

**Comment by clokep** [comment]
🤦 I fixed the epoll issue for asyncio, but not Twisted. Anyway...I think the current version will work?

File: synapse/metrics/_reactor_metrics.py [hunk 1]
```diff
@@ -34,52 +62,94 @@
...
+        elif isinstance(selector, _PollLikeSelector):
+            selector._selector = ObjWrapper(selector._selector, "poll")  # type: ignore[attr-defined]
+            wrapper = selector._selector._wrapped_method  # type: ignore[attr-defined]
+
+        elif isinstance(selector, KqueueSelector):
+            selector._selector = ObjWrapper(selector._selector, "control")  # type: ignore[attr-defined]
+            wrapper = selector._selector._wrapped_method  # type: ignore[attr-defined]
+
+        # XXX Will not work on (Windows-only) ProactorEventLoop.
+    except AttributeError:
```
Comment (clokep) [inline synapse/metrics/_reactor_metrics.py:151]: Maybe this should be wider and have this entire block covered?

**Comment by erikjohnston** [comment]
It seems to work happily on jki.re FWIW

**Comment by clokep** [comment]
> It seems to work happily on jki.re FWIW

With asyncio or just in general?

**Comment by erikjohnston** [comment]
Sorry, that is using asyncio

File: synapse/metrics/_reactor_metrics.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/metrics/_reactor_metrics.py:151]: Yes, I think we should have a very broad `try/except` here, that also includes access to underscored variables. I'd also be tempted to make this an `except Exception`

**Comment by clokep** [comment]
I think this is a bit safer now and gives decent warnings like:

> Skipping configuring ReactorLastSeenMetric: unexpected asyncio loop selector: <selectors.KqueueSelector object at 0x10de4ea50> via <_UnixSelectorEventLoop running=False closed=False debug=False>

or

> Configuring ReactorLastSeenMetric failed: AttributeError("'select.kqueue' object attribute 'control' is read-only")

(the second one was by me introducing a bug intro the code)

**Review by erikjohnston — APPROVED** [review]

**Pull request merged**